spec/concepts/class_member/create_spec.rb

```diff
@@ -5,61 +5,138 @@
-    expect(response[:class_member].school_class).to eq(school_class)
+    response = described_class.call(school_class:, students:)
+    expect(response[:class_members]).to all(have_attributes(school_class:))
   end
 
   it 'assigns the student_id' do
-    response = described_class.call(school_class:, class_member_params:)
-    expect(response[:class_member].student_id).to eq(student.id)
+    response = described_class.call(school_class:, students:)
+    expect(response[:class_members].map(&:student_id)).to match_array(student_ids)
   end
 
-  context 'when creation fails' do
-    let(:class_member_params) { {} }
-
+  context 'when creations fail' do
     before do
       allow(Sentry).to receive(:capture_exception)
     end
 
-    it 'does not create a class member' do
-      expect { described_class.call(school_class:, class_member_params:) }.not_to change(ClassMember, :count)
-    end
+    context 'with malformed students' do
+      let(:students) { nil }
 
-    it 'returns a failed operation response' do
-      response = described_class.call(school_class:, class_member_params:)
-      expect(response.failure?).to be(true)
-    end
+      it 'does not create a class member' do
+        expect { described_class.call(school_class:, students:) }.not_to change(ClassMember, :count)
+      end
 
-    it 'returns the error message in the operation response' do
-      response = described_class.call(school_class:, class_member_params:)
-      expect(response[:error]).to match(/Error creating class member/)
+      it 'returns a failed operation response' do
+        response = described_class.call(school_class:, students:)
+        expect(response.failure?).to be(true)
+      end
+
+      it 'returns the error message in the operation response' do
+        response = described_class.call(school_class:, students:)
+        expect(response[:error]).to match(/No valid students provided/)
+      end
+
+      it 'sent the exception to Sentry' do
+        described_class.call(school_class:, students:)
+        expect(Sentry).to have_received(:capture_exception).with(kind_of(StandardError))
+      end
     end
 
-    it 'sent the exception to Sentry' do
-      described_class.call(school_class:, class_member_params:)
-      expect(Sentry).to have_received(:capture_exception).with(kind_of(StandardError))
+    context 'with a student from a different school' do
+      let(:different_school) { create(:school) }
+      let(:different_school_student) { create(:student, school: different_school) }
+
+      context 'with non existent students' do
+        let(:students) { [different_school_student] }
+
+        it 'does not create a class member' do
+          expect { described_class.call(school_class:, students:) }.not_to change(ClassMember, :count)
+        end
+
+        it 'returns a successful operation response' do
+          response = described_class.call(school_class:, students:)
+          expect(response.success?).to be(true)
+        end
```

Comment:
Why a successful response here?

Reply:
Is it so we can surface the errors from each student?

Reply:
yes its exactly that. So rails will pick up if there's a `response.error` however not `response.errors` so I'm adding to the former for complete request failures but given its a batch endpoint I'm adding to `errors` for the individual cases that error. A little implicit but I think it makes sense and in combination with an empty `class_members` array and `errors` being returned we should be able to surface and expand on this on the frontend.

Let me know what you think though 👍

Reply:
I think that makes sense, let's go with that, and we can always change it if it surfaces some weird edge cases 👍